src/zenml/zen_stores/sql_zen_store.py

```diff
@@ -11449,119 +11557,327 @@ def update_tag(
-            EntityExistsError: If a tag resource relationship with the given
-                configuration already exists.
+            ValueError: If an exclusive tag is being attached to multiple resources
+                of the same type within the same scope.
         """
-        self._set_request_user_id(request_model=tag_resource, session=session)
-        existing_tag_resource = session.exec(
-            select(TagResourceSchema).where(
-                TagResourceSchema.tag_id == tag_resource.tag_id,
-                TagResourceSchema.resource_id == tag_resource.resource_id,
-                TagResourceSchema.resource_type
-                == tag_resource.resource_type.value,
-            )
-        ).first()
-        if existing_tag_resource is not None:
-            raise EntityExistsError(
-                f"Unable to create a tag "
-                f"{tag_resource.resource_type.name.lower()} "
-                f"relationship with IDs "
-                f"`{tag_resource.tag_id}`|`{tag_resource.resource_id}`. "
-                "This relationship already exists."
+        tag_resource_schemas = []
+        for tag_schema, resource_type, resource in tag_resources:
+            try:
+                existing_schema = self._get_tag_resource_schema(
+                    tag_resource=TagResourceRequest(
+                        tag_id=tag_schema.id,
+                        resource_id=resource.id,
+                        resource_type=resource_type,
+                    ),
+                    session=session,
+                )
+            except KeyError:
+                pass
+            else:
+                logger.warning(
+                    f"Tag `{tag_schema.name}` is already assigned to "
+                    f"{resource_type.value} with ID "
+                    f"`{resource.id}`."
+                )
+                tag_resource_schemas.append(existing_schema)
+                continue
+
+            tag_resource_schema = TagResourceSchema.from_request(
+                TagResourceRequest(
+                    tag_id=tag_schema.id,
+                    resource_id=resource.id,
+                    resource_type=resource_type,
+                )
             )
+            session.add(tag_resource_schema)
+            tag_resource_schemas.append(tag_resource_schema)
+
+            # If the tag is an exclusive tag, apply the check and attach/detach accordingly
+            if tag_schema.exclusive:
+                scope_ids: Dict[
+                    TaggableResourceTypes, List[Union[UUID, int]]
+                ] = defaultdict(list)
+                detach_resources: List[TagResourceRequest] = []
+
+                if isinstance(resource, PipelineRunSchema):
+                    if resource.pipeline_id:
+                        scope_ids[TaggableResourceTypes.PIPELINE_RUN].append(
+                            resource.pipeline_id
+                        )
+                    other_runs_with_same_tag = self.list_runs(
```

Comment:
This only returns a page, not all runs. But I assume that's not an issue as there should (in theory) only be one item anyway? In general, not the best for performance to fetch entire runs if we just need the ID I assume.

Reply:
Exactly. For the time being, I will leave this as is. We can optimize it in the future.

Reply:
Optimize it in the future is usually never, and part of the reason why the slowness builds up. Anytime someone uses an exclusive tag on a pipeline run (which hopefully will be a lot if this is a feature that people are using), this will run a highly unoptimized query. Not blocking the release for now, but not a fan either.